### PR TITLE
Implements C++ API for Bifrost

### PR DESCRIPTION
**Pull request by samhorsfield96**
Updates to command-line running of unitig-caller:
- Removes --build function, bifrost graphs built implicitly by Bifrost API
- Implements --call, which returns unitigs within existing graph
- Updates input --refs format for --simple, now takes same format as --call and --query
- Updates unitigs input format for --query and --simple, do now take both fasta and txt file with one unitig per line (with header)
- Removes graph pruning and inexact kmer matching functionality to simplify running

Updates to implementation:
- Updates seqan3 implementation from 3.0.0 to 3.0.2 (conda version)
- Implements Bifrost API, replacing system calls for Bifrost

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,45 @@
 cmake_minimum_required(VERSION 3.4)
 project(map_strings)
 set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -lstdc++fs -fconcepts -march=native")
```
Comment (johnlees) [inline CMakeLists.txt]: Is `-lstdc++fs` still needed here? Also, it should be in target_link_libraries() if it is needed

File: README.md [hunk 2]
```diff
@@ -55,69 +52,77 @@ python setup.py install
...
-to those specified in --input1 (e.g. if 'refs.txt' is given in `--refs`, then 'reads.txt' will
-be given in `--reads` and vice versa), each on new line. No header row.
+#### If pre-built Bifrost graph exists
+
+```
+unitig-caller --call --graph graph.gfa --colours graph.bfg_colors --out out_prefix
+```
 
-`--output` is the prefix for output files.
+`--graph` is a pre-built bifrost graph .gfa, and `--colours` is it's associated colours file.
```
Comment (johnlees) [inline README.md]:
```suggestion
`--graph` is a pre-built bifrost graph .gfa, and `--colours` is its associated colours file.
```

File: README.md [hunk 3]
```diff
@@ -55,69 +52,77 @@ python setup.py install
...
+`--out` is the prefix for output files.
 
-Build mode automatically generates a .fasta file containing unitigs found within the graph.
+Call mode automatically generates a .pyseer file containing unitigs found within the graph and their graph. Rtab or pyseer
+formats can be specified with `--rtab` and `--pyseer` respectively.
 
 ### Running Query mode
-Before running Query mode, generate a coloured compact de Bruijn graph using Build mode.
-Then run the Query command as below.
+Queries existing unitigs in a Bifrost graph.
```
Comment (johnlees) [inline README.md]:
```suggestion
Queries existing unitigs in a Bifrost graph. This is useful when identical unitig definitions need to be used between populations, for example when using pyseer's prediction mode.
```

File: README.md [hunk 4]
```diff
@@ -55,69 +52,77 @@ python setup.py install
...
+
+#### If no pre-built Bifrost graph exists
 ```
-unitig-caller --query --graph-prefix in_prefix --unitigs query_unitigs.fasta --output out_prefix
+unitig-caller --query --refs refs.txt --reads reads.txt --unitigs query_unitigs.fasta --out out_prefix
 ```
 
-`--graph-prefix` is the required prefix for the .gfa, .bfg_colors and unitigs .fasta files generated from
-`--build` mode applied to the new population.
+`refs` and `reads` are the same arguments as in `call`.
```
Comment (johnlees) [inline README.md]:
```suggestion
`--refs` and `--reads` are the same arguments as in `--call`.
```

File: src/bifrost.cpp [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    opt.prefixFilenameOut = output_prefix;
+
+    std::string filename;
+    if (is_ref && (infile_2 == "NA")) {
+        while (std::getline(infile1, filename))
+        {
+            opt.filename_ref_in.push_back(filename);
+        }
+    } else if (!is_ref && (infile_2 == "NA"))
+    {
```
Comment (johnlees) [inline src/bifrost.cpp]:
```suggestion
    } else if (!is_ref && (infile_2 == "NA")) {

```
You might want to try running `clang-format` or similar on the cpp/hpp files

File: src/bifrost.cpp [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        while (std::getline(infile2, filename))
+        {
+            opt.filename_seq_in.push_back(filename);
+        }
+    }
+
+    ColoredCDBG<> ccdbg(opt.k);
+    ccdbg.buildGraph(opt);
+    ccdbg.simplify(opt.deleteIsolated, opt.clipTips, opt.verbose);
+    ccdbg.buildColors(opt);
```
Comment (johnlees) [inline src/bifrost.cpp:49]: Do any of these take a threads option? Or is that already included in the `opt` struct?

File: src/bifrost.cpp [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        for (it_uc; it_uc != it_uc_end; it_uc++)
+        {
+            if (it_uc.getKmerPosition() == position)
+            {
+                colours_arr[it_uc.getColorID()] = 1;
+            }
+        }
+    }
+    // else, look at only position in unitig given by mapping, ignore position
+    else {
```
Comment (johnlees) [inline src/bifrost.cpp:87]:
```suggestion
    } else {
        // else, look at only position in unitig given by mapping, ignore position
```

File: src/bifrost.cpp [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    {
+        ccdbg.write(opt.prefixFilenameOut, opt.nb_threads, opt.verbose);
+    }
+
+    return ccdbg;
+}
+
+// generate colours for a unitig
+template <class T, class U, bool is_const>
+std::vector<bool> generate_colours(const UnitigMap<DataAccessor<T>, DataStorage<U>, is_const> unitig,
```
Comment (johnlees) [inline src/bifrost.cpp:61]: Should `unitig` be passed by reference?

File: src/bifrost.cpp [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+{
+    std::vector<bool> output_array = array1;
+    for (size_t i = 0; i < array1.size(); i++)
+    {
+        if (array1[i] == 1 && array2[i] == 0)
+        {
+            output_array[i] = 0;
+        }
+    }
+    return output_array;
```
Comment (johnlees) [inline src/bifrost.cpp:108]: Not suggesting that you change this here, but if using `std::vector<bool>` turns out to be a pinch-point in ggcaller you could have a look at bitmagic: http://bitmagic.io/apis.html

File: src/bifrost.cpp [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+        // remove new line characters
+        line.erase(std::remove(line.begin(), line.end(), '\n'),
+                    line.end());
+
+        // erase DNA_sequence
+        DNA_sequence.clear();
+
+        // line may be empty so you *must* ignore blank lines
+        if (line.empty())
+            continue;
```
Comment (johnlees) [inline src/bifrost.cpp]: I think it's better to always use braces

File: src/bifrost.cpp [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+        if (line[0] != '>') {
+            DNA_sequence = line;
+        }
+
+        // add to seq list
+        if (DNA_sequence.size() != 0)
+            seq_list.push_back(DNA_sequence);
```
Comment (johnlees) [inline src/bifrost.cpp]: braces

File: src/bifrost.cpp [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+    // iterate to next kmer iterator
+    it_km++;
+
+    // iterate over forward strand of unitig
+    if (!unitig_current.isEmpty)
+    {
+        for (it_km; it_km != it_km_end; it_km++)
+        {
+            // query the kmer in the graph
+            auto unitig_current = ccdbg.find(it_km->first);
```
Comment (johnlees) [inline src/bifrost.cpp:186]: Does this search check for the reverse complement too?

File: src/bifrost.cpp [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+    // initialise unitig map to return
+    std::unordered_map<std::string, std::vector<bool>> unitig_map;
+
+//    // get kmer size
+//    const int kmer = ccdbg.getK();
+
+    // get the number of colours
+    const size_t nb_colours = ccdbg.getNbColors();
+
+    for (const auto um : ccdbg)
```
Comment (johnlees) [inline src/bifrost.cpp]: Iterate over references I think, to avoid a copy

File: src/bifrost.cpp [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+//         for (const auto& i : unitig.second)
+//         {
+//             colours += std::to_string(i);
+//         }
+//         // append to file
+//         outfile << unitig.first << "\t" << colours << "\n";
+//     }
+//
+//     outfile.close();
+// }
```
Comment (johnlees) [inline src/bifrost.cpp]: Remove unused code

File: src/map_bindings.cpp [hunk 15]
```diff
@@ -31,6 +31,199 @@ int py_call_strings(std::vector<std::string> assembly_list,
...
+        input_colour_pref.push_back(p.stem());
+    }
+
+    auto unitig_map = call_unitigs(ccdbg);
+    const std::pair<std::unordered_map<std::string, std::vector<bool>>, std::vector<std::string>> return_pair = std::make_pair(unitig_map, input_colour_pref);
+
+    return return_pair;
+}
+
+std::pair<std::unordered_map<std::string, std::vector<bool>>,
```
Comment (johnlees) [inline src/map_bindings.cpp]: Can you make a `typedef` for this return type (also used above)

File: src/map_bindings.cpp [hunk 16]
```diff
@@ -31,6 +31,199 @@ int py_call_strings(std::vector<std::string> assembly_list,
...
+    }
+
+    // read in compact coloured DBG
+    cout << "Building coloured compacted DBG..." << endl;
+
+    if (infile2 != "NA") {
+        is_ref = 0;
+    }
+
+    ColoredCDBG<> ccdbg;
```
Comment (johnlees) [inline src/map_bindings.cpp]: Is the default constructor needed before running `buildGraph`?

File: src/map_bindings.cpp [hunk 17]
```diff
@@ -31,6 +31,199 @@ int py_call_strings(std::vector<std::string> assembly_list,
...
```
Comment (johnlees) [inline src/map_bindings.cpp]: Necessary?

File: src/map_bindings.cpp [hunk 18]
```diff
@@ -31,6 +31,199 @@ int py_call_strings(std::vector<std::string> assembly_list,
...
+
+    // build graph, write graph if specified
+    size_t lastindex = infile1.find_last_of(".");
+    std::string outgraph = infile1.substr(0, lastindex);
+    ccdbg = buildGraph(infile1, infile2, is_ref, kmer, num_threads, false, write_graph, outgraph);
+
+    std::vector<std::string> query_list = parse_fasta(query_file);
```
Comment (johnlees) [inline src/map_bindings.cpp]: Code looks similar to the function two above – could they be combined, and then make a simple api call function which uses the right one?

File: src/map_bindings.cpp [hunk 19]
```diff
@@ -31,6 +31,199 @@ int py_call_strings(std::vector<std::string> assembly_list,
...
+        std::filesystem::path p(file);
+        input_colour_pref.push_back(p.stem());
+    }
+
+    const std::pair<std::unordered_map<std::string, std::vector<bool>>, std::vector<std::string>> return_pair = std::make_pair(query_colours_map, input_colour_pref);
+
+    return return_pair;
+}
+
 PYBIND11_MODULE(map_strings, m)
```
Comment (johnlees) [inline src/map_bindings.cpp]: We should probably give the extension a more meaningful name. Perhaps `unitig_query`?

File: src/map_bindings.cpp [hunk 20]
```diff
@@ -42,4 +235,32 @@ PYBIND11_MODULE(map_strings, m)
...
+
+  m.def("query_unitigs_build", &py_uc_query_build, "Builds and then queries unitigs in Bifrost graph",
+          py::arg("infile1"),
+          py::arg("kmer"),
+          py::arg("query_file"),
+          py::arg("num_threads") = 1,
+          py::arg("is_ref") = 1,
+          py::arg("write_graph") = 0,
+          py::arg("infile2") = "NA");
+}
```
Comment (johnlees) [inline src/map_bindings.cpp:188]: Can you add the version as an attribute:
```suggestion

    m.attr("version") = VERSION_INFO;
}
```
VERSION_INFO should be set up in setup.py

File: src/map_strings.hpp [hunk 21]
```diff
@@ -28,18 +28,33 @@
...
+        16,
+        10000000,
+        sdsl::sa_order_sa_sampling<>,
+        sdsl::isa_sampling<>,
+        sdsl::plain_byte_alphabet>;
+typedef seqan3::fm_index<seqan3::dna5, seqan3::text_layout::collection, cust_sdsl_wt_index_type> fasta_fm_index;
+using seqan3::operator""_dna5;
 
 // Constants
 const std::string VERSION = "1.1.0";
```
Comment (johnlees) [inline src/map_strings.hpp]: We should remove this. Can you check where it's used (if it is used) and replace with `VERSION_INFO `

File: unitig_caller/__main__.py [hunk 22]
```diff
@@ -120,89 +98,91 @@ def get_options():
...
-        # Read files with prefix input1 as gfa, colours file and fasta file if input2 not specified,
-        # or if input2 specified, use this file for unitig querying. Call `Bifrost query`
+            sys.stderr.write("Calling unitigs within input genomes...\n")
+            unitig_map, input_colour_pref = map_strings.call_unitigs_existing(options.graph, options.colours, options.threads)
 
-        graph_file = options.graph_prefix + ".gfa"
-        colour_file = options.graph_prefix + ".bfg_colors"
-        tsv_file = options.output + ".tsv"
-        rtab_file = options.output + ".rtab"
+        elif options.call and (options.refs != None or options.reads != None) and options.graph == None and options.colours == None:
```
Comment (johnlees) [inline unitig_caller/__main__.py:108]: Might be easier to add a mutually exclusive group for `--mode` in argparse, like here: https://github.com/johnlees/pp-sketchlib/blob/master/pp_sketch/__main__.py#L85-L103

File: unitig_caller/__main__.py [hunk 23]
```diff
@@ -120,89 +98,91 @@ def get_options():
...
+        if options.rtab:
+            sys.stderr.write("Generating rtab file...\n")
+            rtab_format(unitig_map, input_colour_pref, options.out + ".rtab")
 
     elif options.simple:
         # Read input into lists, as in 'index' and 'call'
 
         if options.refs == None or options.unitigs == None:
-            sys.stderr.write("Please specify a strains-list file as input 1 and unitigs file as input 2\n")
+            sys.stderr.write("Error: Please specify a strains-list file as --refs and unitigs file as --unitigs\n")
```
Comment (johnlees) [inline unitig_caller/__main__.py]:
```suggestion
            sys.stderr.write("Error: Please specify a list of assemblies as --refs and unitigs file as --unitigs\n")
```

**Review by johnlees — CHANGES_REQUESTED** [review]

File: src/bifrost.cpp (hunk 6 above)
Comment (samhorsfield96) [reply src/bifrost.cpp:49]: I believe this is already included in the opt structure, as you pass all other arguments this way.

File: CMakeLists.txt (hunk 1 above)
Comment (samhorsfield96) [reply CMakeLists.txt]: No, not needed, removed.

File: src/map_bindings.cpp (hunk 19 above)
Comment (samhorsfield96) [reply src/map_bindings.cpp]: Changed module name to unitig_query

File: src/bifrost.cpp (hunk 8 above)
Comment (samhorsfield96) [reply src/bifrost.cpp:61]: I did try this but it doesn't work from memory; I believe the UnitigMap itself is already some kind of pointer.

File: src/bifrost.cpp (hunk 12 above)
Comment (samhorsfield96) [reply src/bifrost.cpp:186]: Yes, kmers are treated canonically so in creation of a kmer object, it automatically treats the kmer and its reverse complement as the same thing.

File: src/map_bindings.cpp (hunk 20 above)
Comment (samhorsfield96) [reply src/map_bindings.cpp:188]: Not sure how to implement this in setup.py. I can set version to be that in __init__.py however.

File: unitig_caller/__main__.py (hunk 22 above)
Comment (samhorsfield96) [reply unitig_caller/__main__.py:108]: This is already present, however I added these as it enables incorrect argument combinations to be caught.

File: src/map_bindings.cpp (hunk 18 above)
Comment (samhorsfield96) [reply src/map_bindings.cpp]: I've merged the two "exists" and two "build" functions into two single functions, which now take a "call" variable to determine whether call or query is conducted.

**Review by johnlees — APPROVED** [review]

**Pull request merged**